ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+            _,
+            MessageType_KtoU_HydrateFile,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyFilePreConvertToFull,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+}
+
+-(void) testWriteFileFull {
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_WRITE_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+}
+
+- (void) testEventsAreIgnored {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot | FileFlags_IsEmpty;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_APPEND_DATA,
+        KAUTH_VNODE_ADD_SUBDIRECTORY,
+        KAUTH_VNODE_WRITE_SECURITY,
+        KAUTH_VNODE_TAKE_OWNERSHIP,
+        KAUTH_VNODE_ACCESS
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        // Verify for File node
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testFileVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(
+            MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+        MockCalls::Clear();
+
+        // Verify for Directory node
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+        MockCalls::Clear();
+    }
+}
+
+- (void) testOpen {
+    // A file that is not yet in the virtual root, should recieve a creation request
+    testFileVnode->attrValues.va_flags = 0;
+    HandleFileOpOperation(
+        nullptr,
+        nullptr,
+        KAUTH_FILEOP_OPEN,
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        reinterpret_cast<uintptr_t>(filePath),
+        0,
+        0);
+    
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyFileCreated,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            _));
+}
```

Comment:
Could we check `nextCallSequenceNumber` to ensure that the above callback was the only callback called (and that it was only called once)?